File: lib/api/models/ai_interaction.dart
// lib/models/ai_interaction.dart

class AIInteraction {
  final int? interactionId;
  final int userId;
  final String prompt;
  final String? response;
  final int? templateId;
  final String? interactionType;
  final DateTime? createdAt;
  final int? processingTime;
  final int? tokensUsed;
  final bool wasSuccessful;

  // Relationships
  final Map<String, dynamic>? user;
  final Map<String, dynamic>? template;

  AIInteraction({
    this.interactionId,
    required this.userId,
    required this.prompt,
    this.response,
    this.templateId,
    this.interactionType,
    this.createdAt,
    this.processingTime,
    this.tokensUsed,
    this.wasSuccessful = true,
    this.user,
    this.template,
  });

  factory AIInteraction.fromJson(Map<String, dynamic> json) {
    return AIInteraction(
      interactionId: json['interaction_id'],
      userId: json['user_id'],
      prompt: json['prompt'],
      response: json['response'],
      templateId: json['template_id'],
      interactionType: json['interaction_type'],
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : null,
      processingTime: json['processing_time'],
      tokensUsed: json['tokens_used'],
      wasSuccessful: json['was_successful'] ?? true,
      user:
          json['user'] != null ? Map<String, dynamic>.from(json['user']) : null,
      template: json['template'] != null
          ? Map<String, dynamic>.from(json['template'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (interactionId != null) 'interaction_id': interactionId,
      'user_id': userId,
      'prompt': prompt,
      if (response != null) 'response': response,
      if (templateId != null) 'template_id': templateId,
      if (interactionType != null) 'interaction_type': interactionType,
      if (createdAt != null) 'created_at': createdAt!.toIso8601String(),
      if (processingTime != null) 'processing_time': processingTime,
      if (tokensUsed != null) 'tokens_used': tokensUsed,
      'was_successful': wasSuccessful,
      if (user != null) 'user': user,
      if (template != null) 'template': template,
    };
  }

  AIInteraction copyWith({
    int? interactionId,
    int? userId,
    String? prompt,
    String? response,
    int? templateId,
    String? interactionType,
    DateTime? createdAt,
    int? processingTime,
    int? tokensUsed,
    bool? wasSuccessful,
    Map<String, dynamic>? user,
    Map<String, dynamic>? template,
  }) {
    return AIInteraction(
      interactionId: interactionId ?? this.interactionId,
      userId: userId ?? this.userId,
      prompt: prompt ?? this.prompt,
      response: response ?? this.response,
      templateId: templateId ?? this.templateId,
      interactionType: interactionType ?? this.interactionType,
      createdAt: createdAt ?? this.createdAt,
      processingTime: processingTime ?? this.processingTime,
      tokensUsed: tokensUsed ?? this.tokensUsed,
      wasSuccessful: wasSuccessful ?? this.wasSuccessful,
      user: user ?? this.user,
      template: template ?? this.template,
    );
  }
}

File: final_project/models/ai_interaction.py
from sqlalchemy import Column, Integer, String, Text, Boolean, DateTime, CheckConstraint, ForeignKey, Index
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from enum import Enum
from final_project.db import Base

# Enum for interaction types
class InteractionType(str, Enum):
    CHAT = "chat"
    SUGGESTION = "suggestion"
    ANALYSIS = "analysis"

# SQLAlchemy model for AI interactions
class AIInteractionDB(Base):
    __tablename__ = "ai_interaction"

    interaction_id = Column(Integer, primary_key=True, autoincrement=True)  # Primary key
    user_id = Column(Integer, ForeignKey("users.user_id"), nullable=False)  # Foreign key to UserDB
    prompt = Column(Text, nullable=False)  # AI prompt (input from user)
    response = Column(Text)  # AI response (output)
    template_id = Column(Integer, ForeignKey("ai_prompt_template.template_id"))  # Foreign key to AIPromptTemplateDB
    interaction_type = Column(String(50))  # Type of interaction (chat, suggestion, analysis)
    created_at = Column(DateTime, server_default=func.now())  # Timestamp of interaction creation
    processing_time = Column(Integer)  # Processing time in milliseconds
    tokens_used = Column(Integer)  # Number of tokens used in the interaction
    was_successful = Column(Boolean, default=True)  # Whether the interaction was successful
    #interaction_type = Column(SQLEnum(InteractionType), nullable=False)

    __table_args__ = (
        CheckConstraint("interaction_type IN ('chat', 'suggestion', 'analysis')", name="check_interaction_type"),  # Valid interaction types
        Index("idx_ai_interaction_user_id", "user_id"),  # Index on user_id
        Index("idx_ai_interaction_template_id", "template_id"),  # Index on template_id
        Index("idx_ai_interaction_interaction_type", "interaction_type"),  # Index on interaction_type
        Index("idx_ai_interaction_created_at", "created_at"),  # Index on created_at
    )

    # Relationships
    user = relationship("UserDB", back_populates="ai_interactions")  # Relationship with UserDB
    template = relationship("AIPromptTemplateDB", back_populates="interactions")  # Relationship with AIPromptTemplateDB
    prompt_template = relationship("AIPromptTemplateDB", back_populates="interactions")


